File: src/engine/parser.ts
import { Token, TokenType } from "./tokens";
import {
  ProgramNode,
  StatementNode,
  ExpressionNode,
  VariableDeclarationNode,
  AssignmentNode,
  PrintStatementNode,
  IfStatementNode,
  WhileStatementNode,
  BinaryExpressionNode,
  LiteralExpressionNode,
  VariableExpressionNode,
} from "./ast";

export class BhojpuriParser {
  private tokens: Token[];
  private current: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  public parse(): ProgramNode {
    const statements: StatementNode[] = [];
    const line = this.peek().line;

    while (!this.isAtEnd()) {
      statements.push(this.parseStatement());
    }

    return {
      kind: "Program",
      line,
      statements,
    };
  }

  private parseStatement(): StatementNode {
    const token = this.peek();

    if (this.match(TokenType.RAKH)) {
      return this.parseVariableDeclaration();
    } else if (this.match(TokenType.DIKHA)) {
      return this.parsePrintStatement();
    } else if (this.match(TokenType.AGAR)) {
      return this.parseIfStatement();
    } else if (this.match(TokenType.JABTAK)) {
      return this.parseWhileStatement();
    } else if (this.check(TokenType.IDENTIFIER) && this.checkNext(TokenType.ASSIGN)) {
      return this.parseAssignment();
    } else {
      throw new Error(`Syntax Error at line ${token.line}: Unexpected token '${token.value}'`);
    }
  }

  private parseVariableDeclaration(): VariableDeclarationNode {
    const line = this.previous().line;
    const nameToken = this.consume(TokenType.IDENTIFIER, `Syntax Error at line ${line}: Expected variable name after 'rakh'`);
    this.consume(TokenType.ASSIGN, `Syntax Error at line ${nameToken.line}: Expected '=' after variable name in declaration`);
    const initializer = this.parseExpression();

    return {
      kind: "VariableDeclaration",
      line,
      identifier: nameToken.value,
      initializer,
    };
  }

  private parseAssignment(): AssignmentNode {
    const nameToken = this.advance(); // IDENTIFIER
    const line = nameToken.line;
    this.consume(TokenType.ASSIGN, `Syntax Error at line ${line}: Expected '=' after variable name`);
    const value = this.parseExpression();

    return {
      kind: "Assignment",
      line,
      identifier: nameToken.value,
      value,
    };
  }

  private parsePrintStatement(): PrintStatementNode {
    const line = this.previous().line;
    const expression = this.parseExpression();

    return {
      kind: "PrintStatement",
      line,
      expression,
    };
  }

  private parseIfStatement(): IfStatementNode {
    const line = this.previous().line;
    const condition = this.parseExpression();

    const thenBranch: StatementNode[] = [];
    const elseBranch: StatementNode[] = [];

    while (!this.check(TokenType.NAHIN) && !this.check(TokenType.KHATAM) && !this.isAtEnd()) {
      thenBranch.push(this.parseStatement());
    }

    if (this.match(TokenType.NAHIN)) {
      while (!this.check(TokenType.KHATAM) && !this.isAtEnd()) {
        elseBranch.push(this.parseStatement());
      }
    }

    if (!this.match(TokenType.KHATAM)) {
      throw new Error(`Syntax Error at line ${this.peek().line}: Missing 'khatam' to close 'agar' block`);
    }

    return {
      kind: "IfStatement",
      line,
      condition,
      thenBranch,
      elseBranch,
    };
  }

  private parseWhileStatement(): WhileStatementNode {
    const line = this.previous().line;
    const condition = this.parseExpression();

    const body: StatementNode[] = [];

    while (!this.check(TokenType.KHATAM) && !this.isAtEnd()) {
      body.push(this.parseStatement());
    }

    if (!this.match(TokenType.KHATAM)) {
      throw new Error(`Syntax Error at line ${this.peek().line}: Missing 'khatam' to close 'jabtak' block`);
    }

    return {
      kind: "WhileStatement",
      line,
      condition,
      body,
    };
  }

  private parseExpression(): ExpressionNode {
    return this.parseComparison();
  }

  private parseComparison(): ExpressionNode {
    let left = this.parseTerm();

    while (this.match(TokenType.LESS, TokenType.LESS_EQUAL, TokenType.GREATER, TokenType.GREATER_EQUAL, TokenType.EQUAL, TokenType.NOT_EQUAL)) {
      const opToken = this.previous();
      const right = this.parseTerm();
      left = {
        kind: "BinaryExpression",
        line: opToken.line,
        left,
        operator: opToken.value,
        right,
      } as BinaryExpressionNode;
    }

    return left;
  }

  private parseTerm(): ExpressionNode {
    let left = this.parseFactor();

    while (this.match(TokenType.PLUS, TokenType.MINUS)) {
      const opToken = this.previous();
      const right = this.parseFactor();
      left = {
        kind: "BinaryExpression",
        line: opToken.line,
        left,
        operator: opToken.value,
        right,
      } as BinaryExpressionNode;
    }

    return left;
  }

  private parseFactor(): ExpressionNode {
    let left = this.parsePrimary();

    while (this.match(TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MODULO)) {
      const opToken = this.previous();
      const right = this.parsePrimary();
      left = {
        kind: "BinaryExpression",
        line: opToken.line,
        left,
        operator: opToken.value,
        right,
      } as BinaryExpressionNode;
    }

    return left;
  }

  private parsePrimary(): ExpressionNode {
    if (this.match(TokenType.INTEGER)) {
      return {
        kind: "LiteralExpression",
        line: this.previous().line,
        value: parseInt(this.previous().value, 10),
      } as LiteralExpressionNode;
    }
    if (this.match(TokenType.STRING)) {
      return {
        kind: "LiteralExpression",
        line: this.previous().line,
        value: this.previous().value,
      } as LiteralExpressionNode;
    }
    if (this.match(TokenType.SAHI)) {
      return {
        kind: "LiteralExpression",
        line: this.previous().line,
        value: true,
      } as LiteralExpressionNode;
    }
    if (this.match(TokenType.GALAT)) {
      return {
        kind: "LiteralExpression",
        line: this.previous().line,
        value: false,
      } as LiteralExpressionNode;
    }
    if (this.match(TokenType.IDENTIFIER)) {
      return {
        kind: "VariableExpression",
        line: this.previous().line,
        name: this.previous().value,
      } as VariableExpressionNode;
    }

    throw new Error(`Syntax Error at line ${this.peek().line}: Unexpected token '${this.peek().value}'`);
  }

  private match(...types: TokenType[]): boolean {
    for (const type of types) {
      if (this.check(type)) {
        this.advance();
        return true;
      }
    }
    return false;
  }

  private check(type: TokenType): boolean {
    if (this.isAtEnd()) return false;
    return this.peek().type === type;
  }

  private checkNext(type: TokenType): boolean {
    if (this.current + 1 >= this.tokens.length) return false;
    return this.tokens[this.current + 1].type === type;
  }

  private advance(): Token {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private isAtEnd(): boolean {
    return this.peek().type === TokenType.EOF;
  }

  private peek(): Token {
    return this.tokens[this.current];
  }

  private previous(): Token {
    return this.tokens[this.current - 1];
  }

  private consume(type: TokenType, message: string): Token {
    if (this.check(type)) return this.advance();
    throw new Error(message);
  }
}

File: backend/src/main/java/com/bhojpurilang/parser/Parser.java
package com.bhojpurilang.parser;

import com.bhojpurilang.ast.*;
import com.bhojpurilang.model.Token;
import com.bhojpurilang.model.TokenType;

import java.util.ArrayList;
import java.util.List;

public class Parser {
    private final List<Token> tokens;
    private int current = 0;

    public Parser(List<Token> tokens) {
        this.tokens = tokens;
    }

    public Program parse() {
        List<Statement> statements = new ArrayList<>();
        int line = peek().getLine();

        while (!isAtEnd()) {
            statements.add(parseStatement());
        }

        return new Program(statements, line);
    }

    private Statement parseStatement() {
        Token token = peek();

        if (match(TokenType.RAKH)) {
            return parseVariableDeclaration();
        } else if (match(TokenType.DIKHA)) {
            return parsePrintStatement();
        } else if (match(TokenType.AGAR)) {
            return parseIfStatement();
        } else if (match(TokenType.JABTAK)) {
            return parseWhileStatement();
        } else if (check(TokenType.IDENTIFIER) && checkNext(TokenType.ASSIGN)) {
            return parseAssignment();
        } else {
            throw new RuntimeException("Syntax Error at line " + token.getLine() + ": Unexpected token '" + token.getValue() + "'");
        }
    }

    private VariableDeclaration parseVariableDeclaration() {
        int line = previous().getLine();
        Token nameToken = consume(TokenType.IDENTIFIER, "Syntax Error at line " + line + ": Expected variable name after 'rakh'");
        consume(TokenType.ASSIGN, "Syntax Error at line " + nameToken.getLine() + ": Expected '=' after variable name in declaration");
        Expression initializer = parseExpression();
        return new VariableDeclaration(nameToken.getValue(), initializer, line);
    }

    private Assignment parseAssignment() {
        Token nameToken = advance(); // IDENTIFIER
        int line = nameToken.getLine();
        consume(TokenType.ASSIGN, "Syntax Error at line " + line + ": Expected '=' after variable name");
        Expression value = parseExpression();
        return new Assignment(nameToken.getValue(), value, line);
    }

    private PrintStatement parsePrintStatement() {
        int line = previous().getLine();
        Expression expr = parseExpression();
        return new PrintStatement(expr, line);
    }

    private IfStatement parseIfStatement() {
        int line = previous().getLine();
        Expression condition = parseExpression();

        List<Statement> thenBranch = new ArrayList<>();
        List<Statement> elseBranch = new ArrayList<>();

        while (!check(TokenType.NAHIN) && !check(TokenType.KHATAM) && !isAtEnd()) {
            thenBranch.add(parseStatement());
        }

        if (match(TokenType.NAHIN)) {
            while (!check(TokenType.KHATAM) && !isAtEnd()) {
                elseBranch.add(parseStatement());
            }
        }

        if (!match(TokenType.KHATAM)) {
            throw new RuntimeException("Syntax Error at line " + peek().getLine() + ": Missing 'khatam' to close 'agar' block");
        }

        return new IfStatement(condition, thenBranch, elseBranch, line);
    }

    private WhileStatement parseWhileStatement() {
        int line = previous().getLine();
        Expression condition = parseExpression();

        List<Statement> body = new ArrayList<>();

        while (!check(TokenType.KHATAM) && !isAtEnd()) {
            body.add(parseStatement());
        }

        if (!match(TokenType.KHATAM)) {
            throw new RuntimeException("Syntax Error at line " + peek().getLine() + ": Missing 'khatam' to close 'jabtak' block");
        }

        return new WhileStatement(condition, body, line);
    }

    private Expression parseExpression() {
        return parseComparison();
    }

    private Expression parseComparison() {
        Expression left = parseTerm();

        while (match(TokenType.LESS, TokenType.LESS_EQUAL, TokenType.GREATER, TokenType.GREATER_EQUAL, TokenType.EQUAL, TokenType.NOT_EQUAL)) {
            Token opToken = previous();
            Expression right = parseTerm();
            left = new BinaryExpression(left, opToken.getValue(), right, opToken.getLine());
        }

        return left;
    }

    private Expression parseTerm() {
        Expression left = parseFactor();

        while (match(TokenType.PLUS, TokenType.MINUS)) {
            Token opToken = previous();
            Expression right = parseFactor();
            left = new BinaryExpression(left, opToken.getValue(), right, opToken.getLine());
        }

        return left;
    }

    private Expression parseFactor() {
        Expression left = parsePrimary();

        while (match(TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MODULO)) {
            Token opToken = previous();
            Expression right = parsePrimary();
            left = new BinaryExpression(left, opToken.getValue(), right, opToken.getLine());
        }

        return left;
    }

    private Expression parsePrimary() {
        if (match(TokenType.INTEGER)) {
            return new LiteralExpression(Long.parseLong(previous().getValue()), previous().getLine());
        }
        if (match(TokenType.STRING)) {
            return new LiteralExpression(previous().getValue(), previous().getLine());
        }
        if (match(TokenType.SAHI)) {
            return new LiteralExpression(true, previous().getLine());
        }
        if (match(TokenType.GALAT)) {
            return new LiteralExpression(false, previous().getLine());
        }
        if (match(TokenType.IDENTIFIER)) {
            return new VariableExpression(previous().getValue(), previous().getLine());
        }

        throw new RuntimeException("Syntax Error at line " + peek().getLine() + ": Unexpected token '" + peek().getValue() + "'");
    }

    private boolean match(TokenType... types) {
        for (TokenType type : types) {
            if (check(type)) {
                advance();
                return true;
            }
        }
        return false;
    }

    private boolean check(TokenType type) {
        if (isAtEnd()) return false;
        return peek().getType() == type;
    }

    private boolean checkNext(TokenType type) {
        if (current + 1 >= tokens.size()) return false;
        return tokens.get(current + 1).getType() == type;
    }

    private Token advance() {
        if (!isAtEnd()) current++;
        return previous();
    }

    private boolean isAtEnd() {
        return peek().getType() == TokenType.EOF;
    }

    private Token peek() {
        return tokens.get(current);
    }

    private Token previous() {
        return tokens.get(current - 1);
    }

    private Token consume(TokenType type, String message) {
        if (check(type)) return advance();
        throw new RuntimeException(message);
    }
}
